File: dkj-fssc22000/scripts/parse_mdr.py
# -*- coding: utf-8 -*-
"""MDR-001 Rev4 xlsx -> tenants/dkj/data/mdr-catalog.json"""
import zipfile, re, json, sys, os
from datetime import date

DEFAULT = r"g:\내 드라이브\01_업무\HACCP 컨설턴트 1\1) 동김제농협가공센터\● 동김제농협_FSSC22000_V6_운영체계구축\15_문서관리\15-01_MDR\MDR-001_Master_Document_Register_Rev4_동김제농협_FSSC22000.xlsx"
OUT = os.path.join(os.path.dirname(__file__), "..", "data", "mdr-catalog.json")

path = sys.argv[1] if len(sys.argv) > 1 else DEFAULT
if not os.path.exists(path):
    print(json.dumps({"error": "not found", "path": path}, ensure_ascii=False))
    sys.exit(1)

z = zipfile.ZipFile(path)
ss = z.read("xl/sharedStrings.xml").decode("utf-8")
strings = re.findall(r"<t[^>]*>([^<]*)</t>", ss)
wb = z.read("xl/workbook.xml").decode("utf-8")
sheets = re.findall(r'<sheet[^>]+name="([^"]+)"', wb)

# find MDR_Main sheet index
main_idx = sheets.index("MDR_Main") + 1 if "MDR_Main" in sheets else 2
sh = z.read("xl/worksheets/sheet%d.xml" % main_idx).decode("utf-8")

row_map = {}
for m in re.finditer(
    r'<c r="([A-Z]+)(\d+)"([^>]*)>(?:<v>([^<]*)</v>|<is><t>([^<]*)</t></is>)',
    sh,
):
    col, row, attrs, v, inline = m.groups()
    val = inline if inline else v
    if attrs and 't="s"' in attrs and val:
        try:
            val = strings[int(val)]
        except (ValueError, IndexError):
            pass
    if val is None:
        continue
    val = str(val).strip()
    if not val:
        continue
    r = int(row)
    row_map.setdefault(r, {})[col] = val

# MDR_Main columns (from sample row 3):
# A=no B=code C=title D=type E=rev F=제정 G=개정 H=시행 I~L=결재 M=배포 N=위치 O=보존 P=권한 Q=표준 R=관련 S=상태 T=파일경로 U=비고

TYPE_MAP = {
    "PR": "procedure",
    "PRC": "procedure",
    "WI": "sop",
    "FR": "record",
    "F": "record",
    "MDR": "mdr",
    "DOC": "mdr",
    "MN": "manual",
    "HACCP": "manual",
    "PM": "processmap",
    "MAP": "processmap",
}

entries = []
for r in sorted(row_map.keys()):
    if r < 2:
        continue
    c = row_map[r]
    code = c.get("B", "")
    if not code or code in ("문서번호", "Code", "B"):
        continue
    doc_type = c.get("D", "")
    rev_num = c.get("E", "0")
    rev = "Rev%s" % rev_num if rev_num else "Rev0"
    status = c.get("S", "운영중")
    workflow = "완료" if "운영중" in status else "수정필요" if "개정" in status else "완료"
    cat = TYPE_MAP.get(doc_type.upper() if len(doc_type) <= 4 else doc_type[:2], "other")
    if code.startswith("DKJ-P"):
        cat = "procedure"
    elif code.startswith("DKJ-WI") or code.startswith("WI-"):
        cat = "sop"
    elif code.startswith("DKJ-F") or code.startswith("FR-"):
        cat = "record"
    elif code.startswith("PM-"):
        cat = "processmap"
    elif code in ("MDR-001", "DOC-001"):
        cat = "mdr"

    # normalize FR codes: DKJ-F-014 -> FR-014 for portal
    portal_id = code
    if code.startswith("DKJ-F-"):
        portal_id = code.replace("DKJ-F-", "FR-")
    elif code.startswith("F-"):
        portal_id = "FR-" + code[2:]

    entries.append({
        "mdrRow": r,
        "id": portal_id,
        "code": code,
        "title": c.get("C", ""),
        "rev": rev,
        "docType": doc_type,
        "category": cat,
        "workflowStatus": workflow,
        "status": status,
        "filePath": c.get("T", ""),
        "related": c.get("R", ""),
        "note": c.get("U", ""),
        "enacted": c.get("F", ""),
        "revised": c.get("G", ""),
    })

out = {
    "source": path,
    "mdrCode": "MDR-001",
    "mdrRev": "Rev11" if "Rev11" in path else ("Rev4" if "Rev4" in path else "Rev"),
    "parsedAt": str(date.today()),
    "sheets": sheets,
    "entryCount": len(entries),
    "entries": entries,
}

# prefer filename RevN
_m = re.search(r"(Rev\d+)", os.path.basename(path), re.I)
if _m:
    out["mdrRev"] = _m.group(1)

with open(OUT, "w", encoding="utf-8") as f:
    json.dump(out, f, ensure_ascii=False, indent=2)

print("OK", len(entries), "entries ->", OUT)

# JS bundle for file:// (no fetch CORS)
bundle_out = os.path.join(os.path.dirname(OUT), "..", "js", "mdr-catalog.bundle.js")
with open(OUT, "r", encoding="utf-8") as f:
    raw = f.read()
with open(bundle_out, "w", encoding="utf-8") as f:
    f.write("window.DKJ_MDR_CATALOG=" + raw + ";\n")
print("bundle ->", bundle_out)

File: dkj-fssc22000/scripts/parse-mdr.mjs
#!/usr/bin/env node
/** Parse MDR-001 Rev4 xlsx -> JSON summary */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const MDR_PATH = process.argv[2] || String.raw`g:\내 드라이브\01_업무\HACCP 컨설턴트 1\1) 동김제농협가공센터\● 동김제농협_FSSC22000_V6_운영체계구축\15_문서관리\15-01_MDR\MDR-001_Master_Document_Register_Rev4_동김제농협_FSSC22000.xlsx`;

if (!fs.existsSync(MDR_PATH)) {
  console.error('NOT_FOUND', MDR_PATH);
  process.exit(1);
}

// minimal xlsx read via zip
import { createRequire } from 'module';
// use pure zip parse - no xlsx lib
import { execSync } from 'child_process';

const tmpScript = `
import zipfile, re, json, sys
path = sys.argv[1]
z = zipfile.ZipFile(path)
ss = z.read('xl/sharedStrings.xml').decode('utf-8')
strings = re.findall(r'<t[^>]*>([^<]*)</t>', ss)
wb = z.read('xl/workbook.xml').decode('utf-8')
sheets = re.findall(r'<sheet[^>]+name="([^"]+)"', wb)
rows_out = []
for si in range(1, min(len(sheets)+1, 6)):
    try:
        sh = z.read(f'xl/worksheets/sheet{si}.xml').decode('utf-8')
    except:
        continue
    sheet_name = sheets[si-1] if si-1 < len(sheets) else f'sheet{si}'
    # inline strings and shared
    for m in re.finditer(r'<c r="([A-Z]+)(\d+)"([^>]*)>(?:<v>([^<]*)</v>|<is><t>([^<]*)</t></is>)', sh):
        col, row, attrs, v, inline = m.groups()
        val = inline if inline else v
        if attrs and 't="s"' in attrs and val:
            try:
                val = strings[int(val)]
            except:
                pass
        if val and str(val).strip():
            rows_out.append({'sheet': sheet_name, 'cell': col+row, 'value': str(val).strip()})
print(json.dumps({'sheets': sheets, 'cells': rows_out[:500], 'string_count': len(strings)}, ensure_ascii=False))
`;

import os from 'os';
const pyFile = path.join(os.tmpdir(), 'parse_mdr.py');
fs.writeFileSync(pyFile, tmpScript);
const out = execSync(`python "${pyFile}" "${MDR_PATH}"`, { encoding: 'utf8', maxBuffer: 10 * 1024 * 1024 });
const data = JSON.parse(out);
console.log('sheets:', data.sheets);
console.log('strings:', data.string_count);
// find document codes
const codes = data.cells.filter(c => /^(DKJ-P|DKJ-WI|FR-|MDR|DKJ-H|DKJ-S|DOC-)/.test(c.value) || c.value.includes('절차서') || c.value.includes('Rev'));
console.log('sample codes:', codes.slice(0, 60).map(c => c.sheet + ' ' + c.cell + ' ' + c.value).join('\n'));
